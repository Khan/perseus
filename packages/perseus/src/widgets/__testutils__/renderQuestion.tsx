import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
