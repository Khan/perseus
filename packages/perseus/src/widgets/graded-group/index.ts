export {default} from "./graded-group";
