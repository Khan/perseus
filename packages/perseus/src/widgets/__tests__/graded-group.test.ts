import {describe, beforeEach, it} from "@jest/globals";
import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
// TODO(FEI-3857): Include in jest setup so that we don't need to import it everywhere
import "@testing-library/jest-dom/extend-expect";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {question1} from "../__testdata__/graded-group.testdata";

import {renderQuestion} from "./renderQuestion";

import type {APIOptions} from "../../types";

const checkAnswer = () => {
    // NOTE(jeremy): The graded-group widget does not participate in
    // Renderer grading. So we can't call `renderer.score()` and see that
    // the widget is answered correctly. The only route to check the answer
    // is to use the "Check" button that is embedded _inside_ the widget.
    userEvent.click(screen.getByRole("button", {name: "Check"}));
};

describe("graded-group", () => {
    beforeEach(() => {
        // Suppress message from Link component: "Warning: A future version of
        // React will block javascript: URLs as a security precaution. Use
        // event handlers instead if you can. If you need to generate unsafe
        // HTML try using dangerouslySetInnerHTML instead. React was passed
        // "javascript:void(0)""
        jest.spyOn(console, "error").mockImplementation(() => {});

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it.each([true, false])("should snapshot", (isMobile: boolean) => {
        // Arrange and Act
        const {container} = renderQuestion(question1, {
            isMobile,
        });

        // Assert
        expect(container).toMatchSnapshot(
            `initial render (mobile: ${isMobile.toString()})`,
        );
    });

    describe("on desktop", () => {
        it("should be able to be answered correctly", () => {
            // Arrange
            renderQuestion(question1);

            userEvent.click(screen.getAllByRole("button", {name: "True"})[0]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[1]);
            userEvent.click(screen.getAllByRole("button", {name: "True"})[2]);
            userEvent.click(screen.getAllByRole("button", {name: "True"})[3]);

            // Act
            checkAnswer();

            // Assert
            expect(screen.getByRole("alert", {name: "Correct"})).toBeVisible();
            expect(
                screen.queryByRole("alert", {name: "Incorrect"}),
            ).not.toBeInTheDocument();
        });

        it("should be able to be answered incorrectly", () => {
            // Arrange
            renderQuestion(question1);

            userEvent.click(screen.getAllByRole("button", {name: "False"})[0]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[1]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[2]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[3]);

            // Act
            checkAnswer();

            // Assert
            expect(
                screen.queryByRole("alert", {name: "Correct"}),
            ).not.toBeInTheDocument();
            expect(
                screen.getByRole("alert", {name: "Incorrect"}),
            ).toBeVisible();
        });

        it("should display an error if not fully answered", () => {
            // Arrange
            renderQuestion(question1);

            userEvent.click(screen.getAllByRole("button", {name: "False"})[1]);

            // Act
            checkAnswer();

            // Assert
            expect(
                screen.getByText(
                    "We couldn't grade your answer. Make sure you select something for every row.",
                ),
            ).toBeVisible();
        });

        it("should be able to reveal the hint", () => {
            // Arrange
            renderQuestion(question1);

            // Act
            userEvent.click(screen.getByRole("button", {name: "[Hint]"}));
            jest.runOnlyPendingTimers();

            // Assert
            expect(
                screen.getByText(/Some bacteria synthesize their own fuel/),
            ).toBeVisible();
        });

        it("should be able to hide the hint", () => {
            // Arrange
            renderQuestion(question1);
            userEvent.click(screen.getByRole("button", {name: "[Hint]"}));
            jest.runOnlyPendingTimers();

            // Act
            userEvent.click(screen.getByRole("button", {name: "[Hide hint]"}));

            // Assert
            expect(
                screen.queryByText(/Some bacteria synthesize their own fuel/),
            ).not.toBeInTheDocument();
        });
    });

    // The Graded Group on mobile has some different functionality from
    // desktop, so we have a completely different set of tests.
    // Hint titling is also not exactly the same ("Hint" vs "[Hint]")
    describe("on mobile", () => {
        const apiOptions: APIOptions = {
            isMobile: true,
        };

        it("should be able to be answered correctly", () => {
            // Arrange
            renderQuestion(question1, apiOptions);

            userEvent.click(screen.getAllByRole("button", {name: "True"})[0]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[1]);
            userEvent.click(screen.getAllByRole("button", {name: "True"})[2]);
            userEvent.click(screen.getAllByRole("button", {name: "True"})[3]);
            jest.runOnlyPendingTimers();

            // Act
            checkAnswer();

            // Assert
            expect(screen.getByRole("alert", {name: "Correct!"})).toBeVisible();
            // On mobile, you can't interact with the widget at all once you
            // answer correctly.
            expect(
                screen.queryByRole("button", {name: "Check"}),
            ).not.toBeInTheDocument();
        });

        it("should be able to be answered incorrectly", () => {
            // Arrange
            renderQuestion(question1, apiOptions);

            userEvent.click(screen.getAllByRole("button", {name: "False"})[0]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[1]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[2]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[3]);
            jest.runOnlyPendingTimers();

            // Act
            checkAnswer();

            // Assert
            expect(
                screen.getByRole("button", {name: "Try again"}),
            ).toBeVisible();
            expect(screen.getByText("Keep trying")).toBeVisible();
        });

        it("should let the user retry when checked if not fully answered", () => {
            // Arrange
            renderQuestion(question1, apiOptions);

            userEvent.click(screen.getAllByRole("button", {name: "False"})[0]);
            userEvent.click(screen.getAllByRole("button", {name: "False"})[1]);
            jest.runOnlyPendingTimers();

            checkAnswer();

            // Act
            userEvent.click(screen.getAllByRole("button", {name: "False"})[2]);
            jest.runOnlyPendingTimers();

            // Assert
            expect(screen.getByRole("button", {name: "Check"})).toBeVisible();
        });

        it("should be able to reveal the hint", () => {
            // Arrange
            renderQuestion(question1, apiOptions);

            // Act
            userEvent.click(screen.getByRole("button", {name: "Hint"}));

            // Assert
            expect(
                screen.getByText(/Some bacteria synthesize their own fuel/),
            ).toBeVisible();
        });

        it("should be able to hide the hint", () => {
            // Arrange
            renderQuestion(question1, apiOptions);
            userEvent.click(screen.getByRole("button", {name: "Hint"}));

            // Act
            userEvent.click(screen.getByRole("button", {name: "Hide hint"}));

            // Assert
            expect(
                screen.queryByText(/Some bacteria synthesize their own fuel./),
            ).not.toBeInTheDocument();
        });
    });
});
