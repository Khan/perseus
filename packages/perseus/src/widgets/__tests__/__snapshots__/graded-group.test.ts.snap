// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graded-group should snapshot: initial render (mobile: false) 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph perseus-paragraph-centered"
      data-perseus-paragraph-index="0"
    >
      <hr />
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <h2>
        Check your understanding!
      </h2>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="perseus-graded-group"
          >
            <div
              class="title_1lagzkb"
            >
              Metabolic strategies of bacteria
            </div>
            <div
              class="perseus-renderer perseus-renderer-responsive"
            >
              <div
                class="paragraph"
                data-perseus-paragraph-index="0"
              >
                <ol
                  start="1"
                >
                  <li>
                    <div
                      class="paragraph"
                    >
                      <strong>
                        Which of the following statements about metabolic strategies of bacteria are true?
                      </strong>
                    </div>
                    <div
                      class="paragraph"
                    >
                      <div
                        class="perseus-widget-container widget-nohighlight widget-block"
                      >
                        <div
                          class="categorizer-container "
                        >
                          <table
                            class="categorizer-table mobileTable_1wmcwov"
                          >
                            <thead>
                              <tr>
                                <td
                                  class="emptyHeaderCell_1mttrgd"
                                />
                                <th
                                  class="header_whrfk4"
                                >
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        True
                                      </div>
                                    </div>
                                  </div>
                                </th>
                                <th
                                  class="header_whrfk4"
                                >
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        False
                                      </div>
                                    </div>
                                  </div>
                                </th>
                              </tr>
                            </thead>
                            <tbody>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                         Some bacteria conduct photosynthesis and produce oxygen, much like plants.
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        Bacteria are always autotrophic but they may get energy from either light or chemical sources.
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        Some chemosynthetic bacteria introduce energy and fixed carbon into communities where photosynthesis is not possible (e.g., deep-sea vents).
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        Some bacteria live symbiotically inside of host organisms and provide the host with nutrients.
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <span
                                      class="radioSpan_1mq3xs9"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: middle;"
                                        viewBox="0 0 100 99.944"
                                        width="1.0005603137757144em"
                                      >
                                        <path
                                          d="M50.046 8.322q-8.493 0-16.188 3.306-15.561 6.669-22.173 22.23-3.363 7.695-3.363 16.188t3.306 16.188 8.949 13.281q5.586 5.586 13.281 8.892t16.188 3.306 16.188-3.306 13.281-8.892 8.892-13.281 3.306-16.188-3.306-16.188-8.892-13.281-13.281-8.949q-7.695-3.306-16.188-3.306zm0 91.713q-13.623 0-25.137-6.726t-18.183-18.183q-6.726-11.571-6.726-25.137t6.726-25.08 18.24-18.24 25.08-6.669q13.566 0 25.08 6.726 11.514 6.669 18.24 18.183t6.726 25.137-6.726 25.137-18.24 18.126q-11.514 6.726-25.08 6.726z"
                                          fill="currentColor"
                                        />
                                      </svg>
                                    </span>
                                  </div>
                                </td>
                              </tr>
                            </tbody>
                          </table>
                        </div>
                      </div>
                    </div>
                    <div
                      class="paragraph"
                    >
                      <div
                        class="perseus-widget-container widget-nohighlight widget-inline"
                      >
                        <div
                          class="container_167wsvl"
                        >
                          <div
                            class="linkContainer_36rlri-o_O-exerciseLink_574wt1"
                          >
                            <a
                              aria-expanded="false"
                              class="explanationLink_7n4oib"
                              href="javascript:void(0)"
                              role="button"
                            >
                              [Hint]
                            </a>
                          </div>
                        </div>
                      </div>
                    </div>
                  </li>
                </ol>
              </div>
            </div>
            <p
              aria-live="polite"
              role="status"
            />
            <button
              aria-disabled="false"
              class="button_vr44p2-o_O-shared_lwskrm-o_O-default_1w5v67"
              role="button"
              type="button"
            >
              <span
                class="text_f1191h-o_O-LabelLarge_5s82ln-o_O-text_6kf3xs"
              >
                Check
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`graded-group should snapshot: initial render (mobile: true) 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph perseus-paragraph-centered"
      data-perseus-paragraph-index="0"
    >
      <hr />
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <h2>
        Check your understanding!
      </h2>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="gradedGroup_14sgdtb perseus-graded-group"
          >
            <div
              class="title_1lagzkb"
            >
              Metabolic strategies of bacteria
            </div>
            <div
              class="perseus-renderer perseus-renderer-responsive"
            >
              <div
                class="paragraph"
                data-perseus-paragraph-index="0"
              >
                <ol
                  start="1"
                >
                  <li>
                    <div
                      class="paragraph"
                    >
                      <strong>
                        Which of the following statements about metabolic strategies of bacteria are true?
                      </strong>
                    </div>
                    <div
                      class="paragraph"
                    >
                      <div
                        class="perseus-widget-container widget-nohighlight widget-block"
                      >
                        <div
                          class="categorizer-container fullBleedContainer_8ybp6"
                        >
                          <table
                            class="categorizer-table mobileTable_1wmcwov"
                          >
                            <thead>
                              <tr>
                                <td
                                  class="emptyHeaderCell_1mttrgd"
                                />
                                <th
                                  class="header_whrfk4"
                                >
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        True
                                      </div>
                                    </div>
                                  </div>
                                </th>
                                <th
                                  class="header_whrfk4"
                                >
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        False
                                      </div>
                                    </div>
                                  </div>
                                </th>
                              </tr>
                            </thead>
                            <tbody>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                         Some bacteria conduct photosynthesis and produce oxygen, much like plants.
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_0"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_0"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        Bacteria are always autotrophic but they may get energy from either light or chemical sources.
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_1"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_1"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        Some chemosynthetic bacteria introduce energy and fixed carbon into communities where photosynthesis is not possible (e.g., deep-sea vents).
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_2"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_2"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                              </tr>
                              <tr>
                                <td>
                                  <div
                                    class="perseus-renderer perseus-renderer-responsive"
                                  >
                                    <div
                                      class="paragraph"
                                      data-perseus-paragraph-index="0"
                                    >
                                      <div
                                        class="paragraph"
                                      >
                                        Some bacteria live symbiotically inside of host organisms and provide the host with nutrients.
                                      </div>
                                    </div>
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="True"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_3"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                                <td
                                  class="category cell_13aakpm"
                                >
                                  <div
                                    aria-label="False"
                                    class="perseus-interactive"
                                    role="button"
                                  >
                                    <input
                                      class="responsiveInput_1q2zj3l-o_O-responsiveRadioInput_17um91a"
                                      name="perseus_radio_9_3"
                                      type="radio"
                                    />
                                  </div>
                                </td>
                              </tr>
                            </tbody>
                          </table>
                        </div>
                      </div>
                    </div>
                    <div
                      class="paragraph"
                    >
                      <div
                        class="perseus-widget-container widget-nohighlight widget-inline"
                      >
                        <div
                          class="container_167wsvl"
                        >
                          <div
                            class="linkContainer_36rlri"
                          >
                            <a
                              aria-expanded="false"
                              class="mobileExplanationLink_his27d"
                              href="javascript:void(0)"
                              role="button"
                            >
                              Hint
                            </a>
                          </div>
                        </div>
                      </div>
                    </div>
                  </li>
                </ol>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
